### PR TITLE
Add static shim test

### PR DESCRIPTION
**Pull request by jkotalik**
For #1414

### CODE REVIEW

**Comment by jkotalik** [comment]
@dotnet-bot test ci

**Comment by jkotalik** [comment]
@dotnet-bot test CI please

**Comment by jkotalik** [comment]
@dotnet-bot test CI please

**Comment by jkotalik** [comment]
@dotnet-bot test CI please

**Comment by jkotalik** [comment]
@dotnet-bot test CI please

**Comment by pakrym** [comment]
@dotnet-bot test CI please

**Comment by jkotalik** [comment]
@pakrym 
```
09:14:10   
09:14:10   Starting test execution, please wait...
09:14:13   [xUnit.net 00:00:00.55]     Microsoft.AspNetCore.Server.IIS.FunctionalTests.AppOfflineTests.AppOfflineDroppedWhileSiteFailedToStartInRequestHandler_SiteStops_InProcess [SKIP]
09:14:14   Skipped  Microsoft.AspNetCore.Server.IIS.FunctionalTests.AppOfflineTests.AppOfflineDroppedWhileSiteFailedToStartInRequestHandler_SiteStops_InProcess
09:14:26   [xUnit.net 00:00:13.20]     Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.ClientDisconnectTests.ReaderThrowsCancelledException [FAIL]
09:14:26   Failed   Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.ClientDisconnectTests.ReaderThrowsCancelledException
09:14:26   Error Message:
09:14:26    Assert.DoesNotContain() Failure
09:14:26   Found:    (filter expression)
09:14:26   In value: ConcurrentQueue<WriteContext> [WriteContext { EventId = 0, Exception = null, Formatter = Func`3 { ... }, LoggerName = "Microsoft.AspNetCore.Server.IISIntegration.Functio"..., LogLevel = Debug, ... }, WriteContext { EventId = 3, Exception = null, Formatter = Func`3 { ... }, LoggerName = "Microsoft.AspNetCore.Hosting.Internal.WebHost", LogLevel = Debug, ... }, WriteContext { EventId = 4, Exception = null, Formatter = Func`3 { ... }, LoggerName = "Microsoft.AspNetCore.Hosting.Internal.WebHost", LogLevel = Debug, ... }, WriteContext { EventId = 0, Exception = null, Formatter = Func`3 { ... }, LoggerName = "Microsoft.AspNetCore.Hosting.Internal.WebHost", LogLevel = Debug, ... }, WriteContext { EventId = 1, Exception = null, Formatter = Func`3 { ... }, LoggerName = "Microsoft.AspNetCore.Hosting.Internal.WebHost", LogLevel = Information, ... }, ...]
09:14:26   Stack Trace:
09:14:26      at Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.StrictTestServerTests.Dispose() in /_/test/IIS.Tests/StrictTestServerTests.cs:line 16
09:14:26      at ReflectionAbstractionExtensions.DisposeTestClass(ITest test, Object testClass, IMessageBus messageBus, ExecutionTimer timer, CancellationTokenSource cancellationTokenSource) in C:\Dev\xunit\xunit\src\xunit.execution\Extensions\ReflectionAbstractionExtensions.cs:line 76
09:14:26   Standard Output Messages:
09:14:26    | [0.002s] TestLifetime Information: Starting test ReaderThrowsCancelledException at 2018-10-17T16:14:14
09:14:26    | [8.446s] Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.TestServer Debug: Method: GET, RequestUri: 'http://localhost:65284/start', Version: 2.0, Content: <null>, Headers:
09:14:26    |                                                                                       {
09:14:26    |                                                                                       }
09:14:26    | [8.458s] Microsoft.AspNetCore.Hosting.Internal.WebHost Debug: Hosting starting
09:14:26    | [8.460s] Microsoft.AspNetCore.Hosting.Internal.WebHost Debug: Hosting started
09:14:26    | [8.460s] Microsoft.AspNetCore.Hosting.Internal.WebHost Debug: Loaded hosting startup assembly IIS.Tests, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60
09:14:26    | [8.461s] Microsoft.AspNetCore.Hosting.Internal.WebHost Information: Request starting HTTP/1.1 GET http://localhost:65284/start  
09:14:26    | [8.462s] Microsoft.AspNetCore.Hosting.Internal.WebHost Information: Request finished in 1.3202ms 200 
09:14:26    | [8.463s] Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.TestServer Debug: StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnection+HttpConnectionResponseContent, Headers:
09:14:26    |                                                                                       {
09:14:26    |                                                                                         Transfer-Encoding: chunked
09:14:26    |                                                                                         Server: Microsoft-IIS/10.0
09:14:26    |                                                                                         Date: Wed, 17 Oct 2018 16:14:22 GMT
09:14:26    |                                                                                       }
09:14:26    | [8.463s] Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.TestServer Debug: Done
09:14:26    | [8.468s] Microsoft.AspNetCore.Hosting.Internal.WebHost Information: Request starting HTTP/1.1 POST http://localhost/  1
09:14:26    | [8.472s] Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer Debug: Connection ID "17582052951159996793" disconnecting.
09:14:26    | [8.472s] Microsoft.AspNetCore.Hosting.Internal.WebHost Debug: Hosting shutdown
09:14:26    | [8.476s] Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer Debug: Connection ID "17582052951159996793" disconnecting.
09:14:26    | [8.476s] Microsoft.AspNetCore.Hosting.Internal.WebHost Information: Request finished in 7.9937ms 200 
09:14:26    | [8.478s] Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer Error: Unexpected exception in "IISHttpContext.ReadBody".
09:14:26    | System.IO.IOException: IO stopped
09:14:26    |    at Microsoft.AspNetCore.Server.IIS.Core.IO.AsyncIOEngine.Run(AsyncIOOperation ioOperation) in /_/src/Microsoft.AspNetCore.Server.IIS/Core/IO/AsyncIOEngine.cs:line 58
09:14:26    |    at Microsoft.AspNetCore.Server.IIS.Core.IO.AsyncIOEngine.ReadAsync(Memory`1 memory) in /_/src/Microsoft.AspNetCore.Server.IIS/Core/IO/AsyncIOEngine.cs:line 37
09:14:26    |    at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext.ReadBody() in /_/src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.IO.cs:line 95
09:14:26    | [10.679s] TestLifetime Information: Finished test ReaderThrowsCancelledException in 10.6774175s
09:14:26   
```

**Comment by pakrym** [comment]
File it

**Comment by jkotalik** [comment]
🆙 📅

File: IISIntegration.NoV1.sln [hunk 1]
```diff
@@ -36,9 +36,9 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "build", "build", "{7E80C58E
 	ProjectSection(SolutionItems) = preProject
 		build\applicationhost.config = build\applicationhost.config
 		build\applicationhost.iis.config = build\applicationhost.iis.config
-		build\build.msbuild = build\build.msbuild
 		build\Build.Settings = build\Build.Settings
 		build\Config.Definitions.Props = build\Config.Definitions.Props
+		build\copy-native-dlls.targets = build\copy-native-dlls.targets
```
Comment (pakrym) [inline IISIntegration.NoV1.sln]: functional-tests.props?

File: samples/NativeIISSample/Properties/launchSettings.json [hunk 2]
```diff
@@ -14,8 +14,8 @@
       "commandLineArgs": "$(IISExpressArguments)",
       "environmentVariables": {
         "IIS_SITE_PATH": "$(MSBuildThisFileDirectory)",
-        "ANCM_PATH": "$(AncmPath)",
-        "ANCMV2_PATH": "$(AncmV2Path)",
+        "ANCM_PATH": "$(AspNetCoreModuleV1ShimDll)",
```
Comment (pakrym) [inline samples/NativeIISSample/Properties/launchSettings.json]: Fix everywhere or revert.

File: test/Common.FunctionalTests/Inprocess/StartupTests.cs [hunk 3]
```diff
@@ -3,9 +3,11 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (pakrym) [inline test/Common.FunctionalTests/Inprocess/StartupTests.cs]: revert

File: test/Common.FunctionalTests/OutOfProcess/GlobalVersionTests.cs [hunk 4]
```diff
@@ -243,7 +244,7 @@ private static void CopyFiles(DirectoryInfo source, DirectoryInfo target)
             foreach (FileInfo fileInfo in source.GetFiles())
             {
                 var destFileName = Path.Combine(target.FullName, fileInfo.Name);
-                fileInfo.CopyTo(destFileName);
+                fileInfo.CopyTo(destFileName, true);
```
Comment (pakrym) [inline test/Common.FunctionalTests/OutOfProcess/GlobalVersionTests.cs]: Why is this required?

File: test/Common.Tests/Common.Tests.csproj [hunk 5]
```diff
@@ -1,4 +1,4 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (pakrym) [inline test/Common.Tests/Common.Tests.csproj]: revert?

File: test/IIS.StaticShim.FunctionalTests/StaticShimTests.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+            var arch = iisDeploymentParameters.RuntimeArchitecture == RuntimeArchitecture.x64 ? $@"x64\aspnetcorev2.dll" : $@"x86\aspnetcorev2.dll";
+
+            var fileInfo = FileVersionInfo.GetVersionInfo(Path.Combine(AppContext.BaseDirectory, arch));
+            var response = await deploymentResult.HttpClient.GetAsync("/HelloWorld");
+
+            StopServer();
+
+            // File version is sent to debug log (hard to read win32 dll file version in c#)
+            Assert.Contains(TestSink.Writes, context => context.Message.Contains(fileInfo.FileVersion));
```
Comment (pakrym) [inline test/IIS.StaticShim.FunctionalTests/StaticShimTests.cs]: Can you use IISTestSiteFixture and look at loaded modules of the process?

File: test/IIS.FunctionalTests/RequiresNewShim.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+using Microsoft.AspNetCore.Testing.xunit;
+
+namespace Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests
+{
+    [AttributeUsage(AttributeTargets.Assembly | AttributeTargets.Class | AttributeTargets.Method)]
+    public sealed class RequiresNewShimAttribute : Attribute, ITestCondition
+    {
+        public bool IsMet => true;
+
+        public string SkipReason => "Test cannot be run with older shim.";
```
Comment (pakrym) [inline test/IIS.FunctionalTests/RequiresNewShim.cs]: Replace skip reason with something more descriptive.

File: test/IIS.StaticShim.FunctionalTests/StaticShimTests.cs (hunk 6 above)
Comment (jkotalik) [reply test/IIS.StaticShim.FunctionalTests/StaticShimTests.cs]: I'm assuming you are referring to the HostProcess on IISTestSiteFixture?

File: IISIntegration.NoV1.sln [hunk 8]
```diff
@@ -113,6 +113,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "IIS.Tests", "test\IIS.Tests
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Common.Tests", "test\Common.Tests\Common.Tests.csproj", "{D17B7B35-5361-4A50-B499-E03E5C3CC095}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "IIS.StaticShim.FunctionalTests", "test\IIS.StaticShim.FunctionalTests\IIS.StaticShim.FunctionalTests.csproj", "{582B07BC-73F4-4689-8557-B039298BD82C}"
```
Comment (pakrym) [inline IISIntegration.NoV1.sln]: I really don't like the name: StaticShim. BackwardsCompatibility.Tests? RegressionTest?

File: test/IIS.StaticShim.FunctionalTests/StaticShimTests.cs (hunk 6 above)
Comment (pakrym) [reply test/IIS.StaticShim.FunctionalTests/StaticShimTests.cs]: Yep

File: build/repo.props [hunk 9]
```diff
@@ -13,6 +13,15 @@
     <ExcludeSolutions Include="$(RepositoryRoot)IISIntegration.NoV1.sln" />
   </ItemGroup>
 
+  <ItemGroup Condition="'$(RunOnlyANCMBackwardsCompatibilityTests)' == 'true'" >
+      <ExcludeFromTest Include="$(RepositoryRoot)test\IIS.FunctionalTests\*.csproj" />
+      <ExcludeFromTest Include="$(RepositoryRoot)test\IISExpress.FunctionalTests\*.csproj" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(RunNonANCMBackwardsCompatibilityTests)' == 'true'" >
```
Comment (pakrym) [inline build/repo.props]: SkipANCMBackwardsCompatibilityTests?

**Review by pakrym — APPROVED** [review]
Couple last nits
